tex/design_and_implementation.tex

```diff
@@ -1 +1,82 @@
-TODO
+CowHub, as opposed to being monolithic, has a number of micro, modular components that serve different purposes and form a complete system in whole. Each and every component is capable of running on its own, and being built and tested independently, and corresponds to a source repository on GitHub. For an overview of all important components, see Figure \ref{fig:structure}.
+
+\begin{sidewaysfigure}
+  \includegraphics[width=\textwidth]{../sketch/structure.pdf}
+  \caption{The structure of CowHub}
+  \label{fig:structure}
+\end{sidewaysfigure}
+
+CowHub employs Amazon Web Services (AWS) for a scalable, robust and stable cloud solution. For this project, the services that are used include EC2, S3, RDS, Lambda and ElastiCache. The use of each service will be explained further in the following report.
+
+The report proceeds with respect to the components of the project.
+
+\subsection{Database Unit}
+The database unit is built on top of Amazon Relational Database Service (RDS) and Amazon Simple Storage Service (S3), and PostgreSQL has been chosen to be the engine of the database service. The database unit aims to provide a simple, scalable and cloud-based database interface for our other services to use. 
+
+RDS has been chosen because it is an easy solution to a resizable and scalable database, and provides features such as duplication. By employing RDS we could focus more on the application that we are building and pay little to no attention to the time-consuming database administrative tasks.
+
+S3 is used to hold the image uploads. It has been chosen because of its simplicity to use, durable infrastructure and more importantly, its integration with other Amazon and CowHub services.
+
+The database is to store the following:
+
+\begin{enumerate}
+	\item The information about farm and farmers. 
+	\item The information about registered cattle, including the herd number, (optionally) the name, the owner, the birthdate, and most importantly the sample image(s) of its muzzle for identification uses.
+	\item The login details of registered farmers. Farmers will need to log in to be able to input and review the cattle data.
+	\item The match results. Every match job will be assigned a job ID and the result will be stored into the database. These match results will be useful in the future as statistics.
+	\item Other data for the system to run smoothly (including session tokens etc.). 
+\end{enumerate}
+
+The database unit is only accessible through the API Gateway\footnote{Images are encoded in Base64 format before transfer (as a string).}.
+
+\subsection{API Gateway}
```

Comment:
API Gateway refers to our API service built on RoR; I have added a footnote denoting that Amazon API Gateway will be referred to as "Amazon API Gateway".